### ISSUE TITLE
RT-DETR-L conversion to ONNX

### ISSUE THREAD

**Issue by Karimashraf01** [ultralytics/ultralytics#7091]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Export

### Bug

Ultralytics YOLOv8.0.123 🚀 Python-3.10.12 torch-2.1.0+cu121 CPU
rt-detr-l summary: 494 layers, 32148140 parameters, 0 gradients

PyTorch: starting from rtdetr-l.pt with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 300, 84), ()) (63.4 MB)

ONNX: starting export with onnx 1.15.0 opset 17...
ONNX: export failure ❌ 4.1s: Unsupported: ONNX export of operator get_pool_ceil_padding, input size not accessible. Please feel free to request support or submit a pull request on PyTorch GitHub: https://github.com/pytorch/pytorch/issues  [Caused by the value 'input.4 defined in (%input.4 : Float(*, *, *, *, strides=[3297312, 103041, 321, 1], requires_grad=0, device=cpu) = onnx::Pad[mode="constant"](%449, %476, %452), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0 # /usr/local/lib/python3.10/dist-packages/ultralytics/nn/modules/block.py:70:0
)' (type 'Tensor') in the TorchScript graph. The containing node has kind 'onnx::Pad'.] 
    (node defined in /usr/local/lib/python3.10/dist-packages/ultralytics/nn/modules/block.py(70): forward
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py(1508): _slow_forward
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py(1527): _call_impl
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py(1518): _wrapped_call_impl
/usr/local/lib/python3.10/dist-packages/ultralytics/nn/tasks.py(82): _predict_once
/usr/local/lib/python3.10/dist-packages/ultralytics/nn/tasks.py(62): predict
/usr/local/lib/python3.10/dist-packages/ultralytics/nn/tasks.py(45): forward
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py(1508): _slow_forward
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py(1527): _call_impl
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py(1518): _wrapped_call_impl
/usr/local/lib/python3.10/dist-packages/torch/jit/_trace.py(124): wrapper
/usr/local/lib/python3.10/dist-packages/torch/jit/_trace.py(133): forward
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py(1527): _call_impl
/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py(1518): _wrapped_call_impl
/usr/local/lib/python3.10/dist-packages/torch/jit/_trace.py(1285): _get_trace_graph
/usr/local/lib/python3.10/dist-packages/torch/onnx/utils.py(915): _trace_and_get_graph_from_model
/usr/local/lib/python3.10/dist-packages/torch/onnx/utils.py(1011): _create_jit_graph
/usr/local/lib/python3.10/dist-packages/torch/onnx/utils.py(1135): _model_to_graph
/usr/local/lib/python3.10/dist-packages/torch/onnx/utils.py(1596): _export
/usr/local/lib/python3.10/dist-packages/torch/onnx/utils.py(516): export
/usr/local/lib/python3.10/dist-packages/ultralytics/yolo/engine/exporter.py(316): export_onnx
/usr/local/lib/python3.10/dist-packages/ultralytics/yolo/engine/exporter.py(112): outer_func
/usr/local/lib/python3.10/dist-packages/ultralytics/yolo/engine/exporter.py(239): __call__
/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py(115): decorate_context
/usr/local/lib/python3.10/dist-packages/ultralytics/vit/rtdetr/model.py(164): export
/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py(115): decorate_context
<ipython-input-4-c994c7857205>(3): <cell line: 3>
/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py(3553): run_code
/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py(3473): run_ast_nodes
/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py(3257): run_cell_async
/usr/local/lib/python3.10/dist-packages/IPython/core/async_helpers.py(78): _pseudo_sync_runner
/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py(3030): _run_cell
/usr/local/lib/python3.10/dist-packages/IPython/core/interactiveshell.py(2975): run_cell
/usr/local/lib/python3.10/dist-packages/ipykernel/zmqshell.py(539): run_cell
/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py(302): do_execute
/usr/local/lib/python3.10/dist-packages/tornado/gen.py(234): wrapper
/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py(539): execute_request
/usr/local/lib/python3.10/dist-packages/tornado/gen.py(234): wrapper
/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py(261): dispatch_shell
/usr/local/lib/python3.10/dist-packages/tornado/gen.py(234): wrapper
/usr/local/lib/python3.10/dist-packages/ipykernel/kernelbase.py(361): process_one
/usr/local/lib/python3.10/dist-packages/tornado/gen.py(786): run
/usr/local/lib/python3.10/dist-packages/tornado/gen.py(825): inner
/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py(738): _run_callback
/usr/local/lib/python3.10/dist-packages/tornado/ioloop.py(685): <lambda>
/usr/lib/python3.10/asyncio/events.py(80): _run
/usr/lib/python3.10/asyncio/base_events.py(1909): _run_once
/usr/lib/python3.10/asyncio/base_events.py(603): run_forever
/usr/local/lib/python3.10/dist-packages/tornado/platform/asyncio.py(195): start
/usr/local/lib/python3.10/dist-packages/ipykernel/kernelapp.py(619): start
/usr/local/lib/python3.10/dist-packages/traitlets/config/application.py(992): launch_instance
/usr/local/lib/python3.10/dist-packages/colab_kernel_launcher.py(37): <module>
/usr/lib/python3.10/runpy.py(86): _run_code
/usr/lib/python3.10/runpy.py(196): _run_module_as_main
)

    Inputs:
        #0: 449 defined in (%449 : Float(*, 32, *, *, strides=[3276800, 102400, 320, 1], requires_grad=0, device=cpu) = onnx::Relu(%input), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0/ultralytics.nn.modules.conv.Conv::stem1/torch.nn.modules.activation.ReLU::act # /usr/local/lib/python3.10/dist-packages/torch/nn/functional.py:1469:0
    )  (type 'Tensor')
        #1: 476 defined in (%476 : Long(8, strides=[1], device=cpu) = onnx::Cast[to=7](%475), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0 # /usr/local/lib/python3.10/dist-packages/ultralytics/nn/modules/block.py:70:0
    )  (type 'Tensor')
        #2: 452 defined in (%452 : NoneType = prim::Constant(), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0
    )  (type 'NoneType')
    Outputs:
        #0: input.4 defined in (%input.4 : Float(*, *, *, *, strides=[3297312, 103041, 321, 1], requires_grad=0, device=cpu) = onnx::Pad[mode="constant"](%449, %476, %452), scope: ultralytics.nn.tasks.DetectionModel::/ultralytics.nn.modules.block.HGStem::model.0 # /usr/local/lib/python3.10/dist-packages/ultralytics/nn/modules/block.py:70:0
    )  (type 'Tensor')
[]

### Environment

Ultralytics YOLOv8.0.123 🚀 Python-3.10.12 torch-2.1.0+cu121 CPU (Goggle COLAB)

### Minimal Reproducible Example

from ultralytics import RTDETR
model = RTDETR("rtdetr-l.pt")
model.export(format='onnx', dynamic=True, batch=1 ,simplify = 	False)

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@Karimashraf01 hello there! We're sorry to hear you're experiencing trouble exporting RT-DETR-L to ONNX. This appears to be related to a known limitation with the ONNX exporter and its support for certain PyTorch operators. The `get_pool_ceil_padding` error you are seeing indicates that the exporter is unable to handle a specific padding operation in the model graph.

For now, if you would like to assist further, please ensure that you have the latest versions of both PyTorch and the `ultralytics` package, as updates may provide fixes to similar issues over time. Should the problem persist even after updating, contributing through a PR, as you've indicated willingness, would certainly be invaluable. Our community strongly benefits from contributors like yourself.

In your PR, it would be helpful to look at possible alternative implementations of the padding layer that are compatible with the current ONNX export capabilities. Remember to test the modifications sufficiently to ensure that model performance is not compromised.

Appreciate your initiative, and thanks for being part of the YOLO and Ultralytics community! 😊 If you need any guidance, consider reviewing the documentation on model export over at the Ultralytics Docs.

**Comment by Karimashraf01** [comment]
> @Karimashraf01 hello there! We're sorry to hear you're experiencing trouble exporting RT-DETR-L to ONNX. This appears to be related to a known limitation with the ONNX exporter and its support for certain PyTorch operators. The `get_pool_ceil_padding` error you are seeing indicates that the exporter is unable to handle a specific padding operation in the model graph.
> 
> For now, if you would like to assist further, please ensure that you have the latest versions of both PyTorch and the `ultralytics` package, as updates may provide fixes to similar issues over time. Should the problem persist even after updating, contributing through a PR, as you've indicated willingness, would certainly be invaluable. Our community strongly benefits from contributors like yourself.
> 
> In your PR, it would be helpful to look at possible alternative implementations of the padding layer that are compatible with the current ONNX export capabilities. Remember to test the modifications sufficiently to ensure that model performance is not compromised.
> 
> Appreciate your initiative, and thanks for being part of the YOLO and Ultralytics community! 😊 If you need any guidance, consider reviewing the documentation on model export over at the Ultralytics Docs.

well , I did the opposite 😅, I downgraded torch to 2.0.1 as I saw in previous issues that people were able to convert it using the older version, the conclusion is that the newer version of torch changed the implementation of `get_pool_ceil_padding`

**Comment by glenn-jocher** [comment]
@Karimashraf01 apologies for the confusion earlier. It's interesting to note the workaround you've found by downgrading PyTorch. It seems that changes in newer versions of PyTorch have indeed led to issues with ONNX export compatibility for certain operations.

Your observation about `get_pool_ceil_padding` can help us track down the specific changes that could be causing this issue. Although not ideal, downgrading can be a temporary solution while we work on a more permanent fix that's compatible with the latest versions.

Thank you for sharing your findings, and your contributions are valuable to the continuous improvement of the repo. If further assistance is needed, please refer to the Ultralytics Docs on model export, and keep us updated on your progress! 🚀

**Comment by Kr-Sheshank** [comment]
@Karimashraf01 did you able to convert rt-detr best.pt file into onnx file?

**Comment by glenn-jocher** [comment]
@Kr-Sheshank, if you're still encountering issues with the conversion of the `rt-detr` model to ONNX, please ensure you're using compatible versions of the necessary libraries and refer to the model export section in the Ultralytics Docs for guidance. If the problem persists, consider opening an issue on the repo for further assistance. 🛠️

**Comment by Karimashraf01** [comment]
> @Kr-Sheshank, if you're still encountering issues with the conversion of the `rt-detr` model to ONNX, please ensure you're using compatible versions of the necessary libraries and refer to the model export section in the Ultralytics Docs for guidance. If the problem persists, consider opening an issue on the repo for further assistance. 🛠️

Thanks, I was able to convert the model with the compatible versions

**Comment by Karimashraf01** [comment]
> @Karimashraf01 did you able to convert rt-detr best.pt file into onnx file?

I didn't need to i used the pretrained weight and converted it successfully

**Comment by glenn-jocher** [comment]
Great to hear you've successfully converted the pretrained weights to ONNX, @Karimashraf01! If you have further questions or need assistance, feel free to reach out. Happy coding! 🎉
